storage/client_raft_test.go

```diff
@@ -813,3 +815,53 @@ func TestRangeDescriptorSnapshotRace(t *testing.T) {
 		}
 	}
 }
+
+// TestRaftAfterRemoveRange verifies multiraft.state remove the remote
+//  node correctly when range was removed from the store.
+func TestRaftAfterRemoveRange(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	mtc := startMultiTestContext(t, 3)
+	defer mtc.Stop()
+
+	// Make the split.
+	splitArgs := adminSplitArgs(proto.KeyMin, []byte("b"), proto.RaftID(1), mtc.stores[0].StoreID())
+	if _, err := mtc.stores[0].ExecuteCmd(context.Background(), &splitArgs); err != nil {
+		t.Fatal(err)
+	}
+
+	raftID := proto.RaftID(2)
+	mtc.replicateRange(raftID, 0, 1, 2)
+
+	mtc.unreplicateRange(raftID, 0, 2)
+	mtc.unreplicateRange(raftID, 0, 1)
+
+	rng, err := mtc.stores[1].GetRange(raftID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// If the remove range action happen before range apply replica config change, the group
+	// will be created again when multiraft received a MsgApp.
+	if err := util.IsTrueWithin(func() bool {
+		return len(rng.Desc().Replicas) == 1
+	}, 1*time.Second); err != nil {
+		t.Fatal(err)
+	}
+
+	// Remove the range from store2.
+	if err := mtc.stores[1].RemoveRange(rng); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := mtc.transport.Send(&multiraft.RaftMessageRequest{
+		GroupID: proto.RaftID(0),
+		Message: raftpb.Message{
+			From: uint64(mtc.stores[2].RaftNodeID()),
+			To:   uint64(mtc.stores[1].RaftNodeID()),
+			Type: raftpb.MsgHeartbeat,
+		}}); err != nil {
+		t.Fatal(err)
+	}
+	// Execute replica change command to ensure multiraft has processed the previous MsgHeartbeat msg.
```

Comment:
@tschottdorf Because there has no Replica on the second Store, so I use a replication to ensure that MultiRaft has processed the heartbeat .